Try: Don't show a clone when dragging.

jasmussen opened the pull request:
As of #22673, the mover control is now a permanent fixture next to the block switcher, and the entire unit is a draggable handle. One thing that effort helped surface, was that the fact that we make a clone of the block when dragging feels like it breaks with some physics — you duplicate the block when you're supposed to move it. This is because you don't actually move the block until you drop it.

This PR removes the clone, and hides the block being lifted, leaving only the draggable handle. Before:

![before](https://user-images.githubusercontent.com/1204802/84143309-a3309180-aa56-11ea-8180-f6de01e9ee48.gif)

After:

![Kapture 2020-06-22 at 15 44 57](https://user-images.githubusercontent.com/1204802/85295574-6cb83500-b4a0-11ea-8118-0a572d084ca3.gif)

The drag and drop experience feels slightly more precise as a result, and less muddy overall. 

_This description was edited as the PR changed while working on it. You can see the edit history if you like._

jasmussen committed:
Try: Don't show a clone when dragging.

This is an experiment.

mtias commented:
Should the dragged block be removed? It's a bit weird to drag it but still see it in its original position.

jasmussen commented:
I think it should, yes, but I wonder if that's a good experience without a big refactor. I'll see what I can do with CSS, and whether that feels good.

jasmussen committed:
Hide the block being moved.

jasmussen commented:
Just for quick testing, I was able to hack together this:

![remove](https://user-images.githubusercontent.com/1204802/84159707-a71bde00-aa6d-11ea-9e37-27f337bd391c.gif)

I actually think this works rather well, even in this extremely hacky state. 

Depending on feedback, we might want to do this, and use the animate component to transform the height of the block being lifted to zero so that the adjacent blocks rearrange themselves in a nice manner.

ZebulanStanphill commented:
That last mockup looks really nice. I think, design-wise, all it needs is a redesigned chip (pretty much exactly like the one at the bottom of the main post) and some animation so the blocks slide into place rather than instantly pop upwards when the block is removed.

ItsJonQ commented:
@jasmussen Whoa! I just tried this out. I do kinda miss seeing the clone, as it provides some context. But boy.. it **feels** much better. Way smoother. Less noise on the screen.

Also! If we were to go in this direction. Your implementation wouldn't be too off! `display: none` is a great way to handle this - elegant and **performant**.

shaunandrews commented:
I was skeptical of this based on the GIFs alone, but trying it out has me sold. I really love how simple it is, and that I no longer have a bunch of text/elements following my pointer around the screen, nor do I have multiple focus states appearing at once.

I'd love to see more work on expanding the dropzone when hovered for a second or two; Something to give me a sense of what will happen when I let go.

draganescu commented:
This is lovely! I wonder why it still shows the block in the navigation block:

<img width="1024" alt="Screenshot 2020-06-10 at 10 49 39" src="https://user-images.githubusercontent.com/107534/84241490-383f9300-ab08-11ea-9ba9-d15608828e40.png">

Other than that it feels very nice dragging that small symbol

jasmussen commented:
> This feels solid. What is remaining for this to be wrapped up for 5.5? I liked the above idea of swapping the drag handle for the block icon and name. Is that it?

Thanks Michael.

What needs to happen is this:

- `components-draggable__clone` needs to not be created in the first place, because it isn't necessary. 
- When the block being lifted, it's currently simply hidden (`display: none;`) — we should (probably) use the `Animate` component to remove this in a way that lets adjacent blocks animate in to take the space available by the now-lifted block.
- The dark square "chip" that shows up under your cursor when you are performing the drag operation should be a `div` somewhere in the markup, which shows (current thinking), Drag handle, Block type, Block name, [like this more or less](https://user-images.githubusercontent.com/1204802/84143599-2520ba80-aa57-11ea-90e5-216de261b89d.png).

I need help making those changes happen, and I've tentatively pinged @ItsJonQ for help if he has time. But I'm happy to work with anyone who has time, I know you all have so much on your plates these days.

Potentially separate & followup PRs:

- The `grabbing` cursor does now show during the drag operation. Best I can tell, we have to tinker with `effectAllowed` for this to work, simply specifying the cursor does not work during the drag operation: https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/effectAllowed
- To [Shaun's](https://github.com/WordPress/gutenberg/pull/23024#issuecomment-641467054) point, we should look at expanding the dropzone when hovered for a second or two. Not so much that the layout shifts around significantly, but just enough to indicate what'll happen if you release the mouse button.

ItsJonQ commented:
> Depending on feedback, we might want to do this, and use the animate component to transform the height of the block being lifted to zero so that the adjacent blocks rearrange themselves in a nice manner.

@jasmussen Haiii! So! I spent some time playing with this.

I have some good news and some bad news :P.

#### 🐻 Bad news

Height animation (especially during drag/sort interactions) is **really hard**. We use `react-spring` for our animations, which is great! However, their library doesn't support `height: auto` handling:

<img width="1102" alt="Screen Shot 2020-06-11 at 1 45 05 PM" src="https://user-images.githubusercontent.com/2322354/84421593-c8a1d480-abe9-11ea-922f-cb91c37f9908.png">

This is made more complicated by how we (currently) render Block/BlockWrapper components.

#### ☀️ Good news

**It's not impossible**.

I've worked on these interactions before (from scratch even). I may be able to help here if that's the effect we want to achieve 😊 

(I'm biased as I always love fluid interactions)

I have a demo of a simple case of this working:

https://d.pr/v/Wb5zHW

Some things that will be needed to make this work:

* Size measuring functions
* Lifecycle functions that can transition between `null`, `0`, and `clientHeight`
* An additional wrapper to handle height manipulation/overflow
* Leveraging `useSpring` from `react-spring`
* Potentially a custom timing function (to help it feel zippier)

---

I started playing around with drag/resort interactions. I don't think collapsing height is a common pattern.

It's typically one similar to Gutenberg's (clone is attached to cursor, and original stays put). Notion appears to have something similar.

Or something with a shifting placeholder dropzone, like this one:
https://clauderic.github.io/react-sortable-hoc/#/basic-configuration/basic-usage?_k=6w0vhk

---

No action required! Just sharing my notes :)

jasmussen commented:
How important do we feel the animations are for this one?

I looked again at this PR today, and it's trivially easy to remove the bit that clones the block. And in doing so, it [fixes this issue](https://github.com/WordPress/gutenberg/pull/23082#issuecomment-647328885) which is rather frustrating. 

The only remaining part of this PR would then be to invoke whatever is causing the surrounding blocks to shuffle around when you lift and drop a block. The same that gets called when a block gets deleted, presumably. @youknowriad, I recall you working on the block animations — do you know what causes the blocks to move around when you delete a block? And can we do that when you "lift" a block in this PR?

youknowriad commented:
@jasmussen For performance reasons, the animation triggers when the "index" of the block changes, this means if you want the animation to trigger when you start dragging, you should actually remove the block from the block list. That's a bit impactful though and may have some unexpected consequences, so I think it's wiser to try separately.

jasmussen committed:
Remove clone, add shadow, polish comments.

jasmussen commented:
I pushed some fixes:

- I removed the clone from the DOM. The wrapper is still there, to hold the handle being dragged. 
- I added a shadow to the dark gray "bit" indicating the dragged item.

I think with this, we could actually merge this one because it's such a better experience. However some followup items would be nice to look at, whether as fixes to this branch, or separate PRs:

- Move the data-URI SVG for the drag handle into the icon component, and output it in the DOM instead of as a CSS background.
- Clean up the draggable clone bits, it probably could be reduced more.
- Explore how to add animation to the blocks when a block gets "lifted", per Riad's feedback in https://github.com/WordPress/gutenberg/pull/23024#issuecomment-647492477.

jasmussen commented:
I updated the description of this post with a new GIF, and marked it ready for review. I also noted some followup items in https://github.com/WordPress/gutenberg/pull/23024#issuecomment-647531527. Let me know which need to be addressed in this PR, vs. in followups, and feel free to push directly to this branch.

paaljoachim commented:
The drag block for dragging a paragraph block could look like this:
![Icon-drag-drop](https://user-images.githubusercontent.com/5323259/85296486-c2411180-b4a1-11ea-8be3-870d677e5caf.jpg)
To show the drag handles in front of the block name.

jasmussen commented:
> The drag block for dragging a paragraph block could look like this:

Yep, good mockup, this was my first thought too. But then I realized that you'll only ever see this drag handle when you're already dragging, giving it less value. Whereas showing the block type icon & label actually indicates the content of what you're dragging as a little symbolic chip!

ItsJonQ commented:
@jasmussen I'm gonna take a looksy into coding up that "Chip" design 😊

ItsJonQ commented:
Ha HA!

![Screen Capture on 2020-06-22 at 14-45-47](https://user-images.githubusercontent.com/2322354/85324217-696c7b80-b497-11ea-8050-a73f91fd10fd.gif)

The solution is hacky. But the existing code was a little hacky to begin with. Overall, I don't think it's too terrible. It definitely achieves the effect we want!

jasmussen commented:
Love it! Definitely worth a shot, and yes the existing solution definitely was hacky.

I still think there's a place to revisit drag and drop in a holistic rethink kind of way, but this seems like a meaningful improvement to ship in the meantime. Thank you!

ItsJonQ commented:
@jasmussen Awesome! I'll clean up what I have and push it up when ready for testing

Commit:
Add DraggableChip experience to block-draggable

ItsJonQ commented:
@jasmussen Ready for testing!

Just a heads up, I updated the dragging/selecting block's UI a bit. Instead of hiding it completely, I adjusted it to have an opacity fade. The reason is because this avoids having jumps in the layout during the drag experience.

Lemme know how it feels!

![Screen Capture on 2020-06-22 at 15-13-39](https://user-images.githubusercontent.com/2322354/85326642-a2a6ea80-b49b-11ea-9e52-429beca30e92.gif)

ItsJonQ commented:
@jasmussen P.S. I looked into the `grabbing` cursor bug.

In general, I think the `<Draggable />` component needs to be redone 🙈 . (I tried `event.dataTransfer`). The event and rendering mechanics aren't ideal. A symptom of it would be `cursor: grabbing`. Write now, the thing works by a combo of `pointer-events: none` and a massive `z-index` number. It's very strange 🤔 .

We don't need to do that for this PR! Just wanted to make a note.

Other noticeable things:

* It needs to be portal'ed
* Position needs to be set via `transform3d()` rather than `top/left` for rendering performance

MichaelArestad commented:
This is looking great! Most tThere's something strange happening with the nav block where the item disappears instead of greying out. I suspect it's unrelated to this PR. You can see it in action here:

![2020-06-22 13 30 49](https://user-images.githubusercontent.com/1123119/85327850-e47c6480-b48c-11ea-90dd-cd0a2b6012c1.gif)

Commit:
Hide dragging block to emulate a "lift" effect

jasmussen commented:
> There's something strange happening with the nav block where the item disappears instead of greying out. I suspect it's unrelated to this PR. 

This was a temporary issue, I believe, that was fixed with https://github.com/WordPress/gutenberg/pull/23024/commits/c38dfa088f30a5066c28578c33b1796632dea238. 

There are actually two big changes going on here:

1. There's no clone shown next to the cursor when you're dragging. Just the dark chip with icon and label.
2. As soon as you _lift_ a block off the page to move it to a new spot, you do literally lift it (the source block is hidden).

Like so:

![move](https://user-images.githubusercontent.com/1204802/85366494-bdbf3c00-b527-11ea-8576-b75865f7aa7c.gif)

_Specifically it highlights an issue with the navigation menu block, that the drag handle is so far from the block itself. This is a separate issue, worth discussing separately._

jasmussen commented:
For those following along, here's a refresher of the PR. Current status is this:

![dragondrop](https://user-images.githubusercontent.com/1204802/85366702-2efeef00-b528-11ea-83ef-93420a4fdd82.gif)

When you perform a drag and drop operation, you **lift a block off the page**, it transforms into a dark chip that indicates the block you're moving, and then it transforms back into a block when you drop it. 

This solves a number of problems:

- It gives more space for the drag operation, as the block being moved is taken out of the flow when lifted.
- It avoids the muddiness that the background-less clone caused ([here's why we can't add a background](https://github.com/WordPress/gutenberg/issues/21391#issuecomment-646017755))
- It scales better to if/when we can add drag and drop to multiple blocks at once.

Currently it feels a little jarring because the page jumps when you lift the block. This should be addressable per [the note here](https://github.com/WordPress/gutenberg/pull/23024#issuecomment-647492477), and I think we should get this in as is and then try and address Riad's point here. If we find that doesn't work, the code is trivial to revisit, but it feels like it's worth trying this out in the plugin for a bit at least.

jasmussen committed:
Only show the grab cursor when drag and drop is possible.

jasmussen commented:
Checks are passing ;) 

@youknowriad can you take this one for a spin and see how it feels, and evaluate your confidence that we can add the "block removed" animation?

youknowriad commented:
This does feel cool, I'd judge at 50% the possibilities to have a successful animation while the block is being dragged 😬. It's not going to be trivial for sure.

jasmussen committed:
Remove dead JS.

jasmussen commented:
Discovered this issue:

![scale issue](https://user-images.githubusercontent.com/1204802/85392505-e822f080-b54b-11ea-8d3d-18e5aa7f1ca8.gif)

and fixed it:

![drag issue](https://user-images.githubusercontent.com/1204802/85392508-ea854a80-b54b-11ea-9c97-04e707098011.gif)

Matías pointed out

youknowriad commented on packages/components/src/draggable/index.js:
```diff
@@ -6,12 +6,11 @@ import { noop } from 'lodash';
  * WordPress dependencies
  */
-import { Component } from '@wordpress/element';
+import { Component, createRef } from '@wordpress/element';
```
There's a lot of changes in this file, I just wanted to mention that this is a generic Draggable component which means it shouldn't be modeled with the block draggable in mind. If we want an option to remove the clonable, it shouldn't be automatic as it breaks backward compatibility.

jasmussen replied:
Hmm, indeed, that may affect this change I made: https://github.com/WordPress/gutenberg/pull/23024/commits/bac44fd5f6cf2caa9fb83f60f48da574a5e78dc1

I'm curious how to best proceed, as I'm getting in the deep end here. Do you have time to help me out with this one?

youknowriad replied:
I'm afraid I won't have time today but I can look tomorrow... ideally you should be able to use Storybook to work on this component  with multiple variations.

ellatrix replied:
Could we just revert to hiding the clone?

youknowriad replied:
What did we end up doing here?

jasmussen commented:
> This does feel cool, I'd judge at 50% the possibilities to have a successful animation while the block is being dragged 😬. It's not going to be trivial for sure.

Was also just reminded that this might affect undo/redo.

Are there any simple approaches that might accomplish this? I am almost certainly _underthinking_ it, but is there any other way to send the signal to animate around, that deleting a block does?

paaljoachim commented:
What I see is dragging a block a blue line shows up in various locations, but not the start location. The blue line should always show up as it gives a hint to where the user can drop the block (also in the default start position where drag began.) As the user might change their mind in relation to dragging a block, and need the visual hint to get it back to where it started from.

jasmussen commented:
> What I see is dragging a block a blue line shows up in various locations, but not the start location. The blue line should always show up as it gives a hint to where the user can drop the block (also in the default start position where drag began.) As the user might change their mind in relation to dragging a block, and need the visual hint to get it back to where it started fro

This was part of the reason for going with the clone concept and the blue lines in the first place, which is in contrast to other forms of drag and drop where you instantly move a block, and adjacent blocks rearrange themselves.

I used to be in the "blue line" camp for exactly the "what if I change my mind" camp, but have realized that we have big and visible undo buttons which means it's trivial to revert a change.

Your idea, showing a blue line in the spot of the block you just "lifted", is something we can try — my instinct is it simply adds confusion, because the blue line is meant to indicate dropzone, not "here's where a block used to be". There might also be a different treatment which does not indicate dropzone, that we can try. But in any case, I think it should be a followup.

ZebulanStanphill commented:
Behavior-wise, I think this is a definite improvement over `master`, and I'd like to see it get in before WP 5.5.

jasmussen commented:
It seems there's some consensus this is worth trying. I obviously agree, and see this path forward:

- Merge after [these fixes](https://github.com/WordPress/gutenberg/pull/23024#discussion_r444121775), try it out.
- Explore a followup that actually removes the block from the position it's lifted, and places it where it's dropped, so that blocks rearrange themselves. This might be similar to how cut and paste works currently, CC: @mcsf if you can support that notion.
- If the followup is complex, and if we don't like the behavior without the animation, it's trivial to unhide the block being dragged again.

ellatrix commented on packages/block-editor/src/components/block-draggable/draggable-chip.js:
```diff
@@ -0,0 +1,70 @@
+		label = blockType.title;
+	}
+
+	return <BlockDraggableChip icon={ icon } label={ label } />;
```
Any reason this is a separate component?

tomtev commented:
Is it an idea to keep the block that is being moved and just add opacity to it? Then the content don't jump so much. 
This is how we have done it with UX Builder:

![Mjvco3dvNJ](https://user-images.githubusercontent.com/48070/85550539-5be2fd00-b621-11ea-8735-103d53bdfb88.gif)

Commit:
Polish chip.

jasmussen commented:
I rebased and polished the chip to look like this:

![handle](https://user-images.githubusercontent.com/1204802/85676531-f9dcd300-b6c6-11ea-8cbf-6a70aaf34258.gif)

What I realized was that the label is not that useful, and although I didn't initially think the handle was useful, it serves well to sit under the cursor that's dragging and covering. 

Per feedback in https://github.com/WordPress/gutenberg/pull/23024#discussion_r444138760, I'm unsure how to proceed. 

If @youknowriad or @ellatrix have time to take a look at this one and land it, feel free to push directly to this branch.

tomtev commented on packages/components/src/draggable/style.scss:
```diff
@@ -23,3 +23,8 @@ body.is-dragging-components-draggable {
+
+// This is only for testing.
+.is-dragging.is-selected {
+	display: none !important;
```
did you try with `opacity: .5` instead here? Would work much better with for example reordering columns.

jasmussen replied:
We tried with ´opacity: 0.7`. 

As I see it, there are two paths forward:

1. Go with the more radical solution of **pulling the block out of the flow**, literally lifting it off the page. This benefits very big blocks which take up a lot of vertical space, but can cause jumpiness.
2. Go with the opacity change, and the chip, which is still an improvement over what we're shipping, but does not have the space saving benefits.

Honestly I think we should get option 1 in good shape, try it in the plugin, and then go with option 2 if we find after using it that the jump is jarring.

ellatrix committed:
Polish DraggableChip

ellatrix committed:
Leave clone logic

jasmussen commented:
Thank you for the help Ella, let's get this in and test it out, and I'll do any follow-ups based on feedback.

mtias commented:
Agreed, let's see how it feels.
